### ISSUE TITLE
confusion_matrix_run

### ISSUE THREAD

**Issue by 66040233122-blip** [SolightzZ/Dataset_learning-#1]
# lab 4
import numpy as np
import pandas as pd

# -----------------------------
# 1) โหลดและเตรียมข้อมูล train
# -----------------------------
file_path = r"Dataset path learning floor matrix task.csv"  # เปลี่ยนเป็น path ของคุณ
data = pd.read_csv(file_path)  # โหลดไฟล์ CSV เข้ามาเป็น DataFrame

# เลือกเฉพาะฟีเจอร์ที่ต้องการใช้ (Peabody, Raven, SAQ)
# และแปลงค่าในคอลัมน์ 'Group' เป็นตัวเลข: 'Down' = 0, 'TD' = 1
X = data[['Peabody', 'Raven', 'SAQ']].values  # ดึงข้อมูล features เป็น numpy array
y = data['Group'].map({'Down': 0, 'TD': 1}).values  # แปลง label เป็นตัวเลข

# รวมข้อมูล features และ labels เข้าด้วยกันในรูปแบบ numpy array 
# โดยมีคอลัมน์ 4 คอลัมน์คือ x1, x2, x3, และ class (label)
train_data = np.column_stack((X, y))

# -----------------------------
# 2) รับข้อมูล test จากผู้ใช้
# -----------------------------
# รับค่าคะแนนจากผู้ใช้เพื่อใช้เป็นข้อมูลทดสอบ (test sample)
x1_test = float(input("กรอกค่า x1 (Peabody) - คะแนนวัด ทักษะด้านภาษา / คำศัพท์ (ช่วง 40–120): "))
x2_test = float(input("กรอกค่า x2 (Raven): - คะแนนวัด การคิดเชิงตรรกะ / เหตุผลเชิงนามธรรม (ช่วง 7-25):"))
x3_test = float(input("กรอกค่า x3 (SAQ): - คะแนนวัด ทักษะทางสังคม (ช่วง 6-18):"))
test_data = np.array([x1_test, x2_test, x3_test])  # สร้าง numpy array ของข้อมูลทดสอบ

# -----------------------------
# 3) คำนวณ Euclidean Distance
# -----------------------------
# คำนวณระยะห่างแบบ Euclidean ระหว่างข้อมูลทดสอบกับข้อมูลฝึกแต่ละตัว
# โดยใช้สูตร sqrt( (x1 - x1_test)^2 + (x2 - x2_test)^2 + (x3 - x3_test)^2 )
distances = np.sqrt(np.sum((train_data[:, :3] - test_data) ** 2, axis=1))

# -----------------------------
# 4) เรียงลำดับระยะห่าง
# -----------------------------
# หา index ของข้อมูลฝึกที่มีระยะห่างจากข้อมูลทดสอบใกล้ที่สุดเป็นอันดับต้น ๆ
indices = np.argsort(distances)  # คืนลำดับ index จากระยะใกล้ไปไกล

# -----------------------------
# 5) เลือก K เพื่อนบ้านที่ใกล้ที่สุด
# -----------------------------
K = 3  # จำนวนเพื่อนบ้านที่ใช้ในการตัดสินใจ สามารถเปลี่ยนค่าได้
nearest_classes = train_data[indices[:K], 3]  # ดึงคลาสของ K เพื่อนบ้านใกล้สุด

# -----------------------------
# 6) นับจำนวนคลาส และหาคลาสที่มากที่สุด
# -----------------------------
# นับจำนวนของแต่ละคลาสใน K เพื่อนบ้าน
class_counts = np.bincount(nearest_classes.astype(int))  # แปลงเป็น int ก่อนนับ
predicted_class = np.argmax(class_counts)  # เลือกคลาสที่มีจำนวนมากที่สุด

# -----------------------------
# 7) แสดงผลลัพธ์
# -----------------------------
print("\nTest Data:", test_data)
if predicted_class == 0:
    print("ผลลัพธ์การทำนาย: Down Syndrome")  # ถ้าค่าที่ได้คือ 0 แปลว่า Down Syndrome
else:
    print("ผลลัพธ์การทำนาย: TD (ปกติ)")  # ถ้าไม่ใช่ 0 คือ TD (ปกติ)

#Lab 5
import numpy as np 
import pandas as pd
from sklearn.model_selection import KFold
from sklearn.metrics import confusion_matrix, accuracy_score, recall_score, precision_score, f1_score
from sklearn.naive_bayes import GaussianNB
from sklearn.tree import DecisionTreeClassifier
from sklearn.neighbors import KNeighborsClassifier
from sklearn.impute import SimpleImputer   # สำหรับจัดการ NaN

# numpy, pandas → จัดการข้อมูล
# KFold → แบ่งข้อมูลสำหรับ cross-validation
# confusion_matrix, accuracy_score,... → คำนวณ metrics การประเมินผล
# GaussianNB → โมเดล Naive Bayes
# DecisionTreeClassifier → โมเดล Decision Tree
# KNeighborsClassifier → โมเดล KNN (k=3)
# SimpleImputer → จัดการค่าที่หาย (NaN)

# -----------------------------
# 1) โหลด Dataset
# -----------------------------
file_path = r"Dataset path learning floor matrix task.csv"  # เปลี่ยนเป็น path ของไฟล์ dataset ของคุณ
data = pd.read_csv(file_path)

# เตรียม features และ labels
X = data[['Peabody', 'Raven', 'SAQ']].values
y = data['Group'].map({'Down': 0, 'TD': 1}).values

# จัดการ NaN (แทนที่ด้วยค่าเฉลี่ยของแต่ละ column)
imputer = SimpleImputer(strategy="mean")
X = imputer.fit_transform(X)

# -----------------------------
# 2) ฟังก์ชัน Cross Validation
# -----------------------------
def run_cross_validation(X, y, model_name="NaiveBayes", n_splits=5, k=3):
    kf = KFold(n_splits=n_splits, shuffle=True, random_state=42)

    acc_list, rec_list, prec_list, f1_list = [], [], [], []
    fold = 1  
    
    for train_index, test_index in kf.split(X):
        print(f"\n================ Fold {fold} ({model_name}) ================")
        
        X_train, X_test = X[train_index], X[test_index]
        y_train, y_test = y[train_index], y[test_index]

        # ---------- เลือกโมเดล ----------
        if model_name == "NaiveBayes":
            model = GaussianNB()
        elif model_name == "DecisionTree":
            model = DecisionTreeClassifier(random_state=42)
        elif model_name == "KNN":
            model = KNeighborsClassifier(n_neighbors=k)
        else:
            raise ValueError("Model name not supported")

        model.fit(X_train, y_train)
        y_pred = model.predict(X_test)

        # ---------- Evaluate ----------
        cm = confusion_matrix(y_test, y_pred, labels=[0,1])
        cm_df = pd.DataFrame(cm, index=["Actual Down", "Actual TD"],
                                columns=["Pred Down", "Pred TD"])
        print("Confusion Matrix:")
        print(cm_df)

        acc = accuracy_score(y_test, y_pred)  * 100
        rec = recall_score(y_test, y_pred, pos_label=1)  * 100
        prec = precision_score(y_test, y_pred, pos_label=1)  * 100
        f1 = f1_score(y_test, y_pred, pos_label=1)  * 100

        print(f"Accuracy : {acc:.2f}%")
        print(f"Recall   : {rec:.2f}%")
        print(f"Precision: {prec:.2f}%")
        print(f"F-Measure: {f1:.2f}%")

        acc_list.append(acc)
        rec_list.append(rec)
        prec_list.append(prec)
        f1_list.append(f1)
        
        fold += 1  

    # -----------------------------
    # Summary ของโมเดลนี้
    # -----------------------------
    print(f"\n=========== Summary {model_name} (Average from {n_splits} folds) ===========")
    print(f"Mean Accuracy : {np.mean(acc_list):.2f}%")
    print(f"Mean Recall   : {np.mean(rec_list):.2f}%")
    print(f"Mean Precision: {np.mean(prec_list):.2f}%")
    print(f"Mean F-Measure: {np.mean(f1_list):.2f}%")

    return np.mean(acc_list), np.mean(rec_list), np.mean(prec_list), np.mean(f1_list)

# -----------------------------
# 3) Run Cross Validation ทั้ง 3 โมเดล
# -----------------------------
results = {}
results["NaiveBayes"] = run_cross_validation(X, y, model_name="NaiveBayes", n_splits=5)
results["DecisionTree"] = run_cross_validation(X, y, model_name="DecisionTree", n_splits=5)
results["KNN"] = run_cross_validation(X, y, model_name="KNN", n_splits=5, k=3)

# -----------------------------
# 4) สรุปผลรวมของทั้ง 3 โมเดล
# -----------------------------
print("\n================ Summary of All Models ================")
df_results = pd.DataFrame(results, index=["Accuracy", "Recall", "Precision", "F-Measure"])
print(df_results)

print("\n=========== Final Average across models ===========")
print(df_results.mean(axis=1))

#dataset

ID,Group,Gender,Age_months,Peabody,Raven,SAQ,PMA-SR-K1,GPT_total,WM_matr_sequential,WM_matr_simultaneous,Floor Matrix Map,Floor Matrix Obs,Counterbalancing floor matrix task
1,TD,F,66,43,25,11,6,18,3,4,3,4,Observ-Map
2,TD,F,64,45,11,12,7,14.00000001,6,6,3,3,Map-Observ
3,TD,M,63,47,18,14,5,14.99999999,5,4,3,4,Observ-Map
4,TD,M,70,47,21,17,8,13,5,4,3,4,Map-Observ
5,TD,M,66,48,20,11,7,16,7,6,4,6,Observ-Map
6,TD,M,69,53,14,11,4,13,3,2,3,3,Map-Observ
7,TD,F,67,55,15,9,6,14.99999999,4,4,3,4,Map-Observ
8,TD,M,66,56,16,6,7,18,4,4,2,4,Map-Observ
9,TD,M,62,56,19,11,9,16,4,3,3,2,Observ-Map
10,TD,F,68,57,11,10,9,18,7,7,3,6,Map-Observ
11,TD,M,69,61,19,15,6,13,4,6,4,5,Observ-Map
12,TD,M,64,64,25,16,9,14.00000001,5,6,4,3,Map-Observ
13,TD,M,65,66,27,8,10,14.99999999,5,6,5,6,Map-Observ
14,TD,M,66,68,19,12,10,17.00000001,6,6,5,6,Map-Observ
15,TD,M,69,72,12,13,5,16,5,2,3,3,Map-Observ
16,TD,M,62,73,15,11,7,14.00000001,6,6,2,3,Observ-Map
17,TD,M,62,74,20,13,8,14.99999999,5,5,4,3,Observ-Map
18,TD,M,65,74,24,14,9,14.99999999,6,5,5,4,Observ-Map
19,TD,M,62,75,18,14,7,14.99999999,6,5,2,4,Map-Observ
20,TD,M,67,78,17,15,6,14.00000001,5,5,3,4,Map-Observ
21,TD,F,63,78,18,13,7,19.00000001,5,3,4,3,Observ-Map
22,TD,F,68,80,20,14,9,18,6,5,4,5,Observ-Map
23,TD,F,65,83,20,9,11,17.00000001,7,7,4,6,Map-Observ
24,TD,M,72,84,22,16,10,16,7,6,4,6,Map-Observ
25,TD,F,67,87,21,10,10,16,6,4,4,4,Observ-Map
26,TD,M,69,87,23,11,7,19.99999999,5,5,4,4,Map-Observ
27,TD,F,66,88,22,11,8,16,5,5,3,4,Observ-Map
28,TD,F,68,89,20,16,11,18,6,6,4,6,Map-Observ
29,TD,F,66,91,19,14,9,19.00000001,6,7,4,5,Map-Observ
30,TD,M,62,95,13,10,6,13,3,5,4,3,Observ-Map
31,Down,F,129,93,7,6,9,4.999999998,3,3,3,3,Observ-Map
32,Down,M,153,33,8,,3,8.000000001,2,1,1,1,Observ-Map
33,Down,F,164,55,8,10,3,11.99999999,4,1,1,3,Map-Observ
34,Down,F,215,115,8,6,6,9.000000009,3,3,2,3,Observ-Map
35,Down,M,158,47,9,12,4,16,3,2,1,2,Map-Observ
36,Down,F,202,84,9,14,2,9.999999996,7,5,2,5,Map-Observ
37,Down,F,98,49,10,12,4,14.00000001,5,4,2,5,Map-Observ
38,Down,M,213,51,10,9,3,13,0,0,0,1,Observ-Map
39,Down,M,199,80,10,10,3,14.00000001,5,0,1,3,Observ-Map
40,Down,M,96,28,11,2,4,6.000000006,2,1,2,2,Observ-Map
41,Down,M,133,79,11,6,5,13,2,3,3,3,Observ-Map
42,Down,M,180,84,11,12,3,13,8,5,2,6,Observ-Map
43,Down,M,,27,12,11,0,14.99999999,4,1,4,3,Observ-Map
44,Down,M,188,11,13,9,3,11.99999999,0,0,1,1,Map-Observ
45,Down,M,102,54,13,8,6,9.000000009,3,3,3,4,Map-Observ
46,Down,M,110,62,13,4,5,6.000000006,2,3,5,4,Observ-Map
47,Down,M,104,73,13,7,5,13,2,2,5,4,Map-Observ
48,Down,F,98,71,14,4,6,9.000000009,1,1,4,4,Map-Observ
49,Down,M,93,37,16,5,4,6.999999993,4,1,3,2,Map-Observ
50,Down,M,159,97,16,9,6,6.999999993,3,3,4,4,Observ-Map
51,Down,F,195,67,18,14,8,14.99999999,8,3,4,5,Observ-Map
52,Down,M,144,88,18,17,8,11,6,3,3,4,Map-Observ
53,Down,M,130,77,19,6,10,9.000000009,2,4,4,4,Map-Observ
54,Down,F,202,111,19,18,5,14.00000001,7,8,5,5,Observ-Map
55,Down,F,119,82,20,8,3,14.99999999,3,2,7,3,Map-Observ
56,Down,F,146,87,20,8,9,4.999999998,1,1,3,4,Map-Observ
57,Down,F,120,77,21,7,8,11,2,4,3,3,Observ-Map
58,Down,M,212,95,22,7,6,13,3,1,2,4,Map-Observ
59,Down,M,165,35,23,8,9,21,6,7,3,5,Map-Observ
60,Down,M,199,67,23,15,9,17.00000001,7,8,7,5,Map-Observ
